### ISSUE TITLE
Non-secure connection to local IP refused

### ISSUE THREAD

**Issue by spkersten** [flutter/flutter#72723]
Making a request over http to a local IP fails with: `Bad state: Insecure HTTP is not allowed by platform`.
This works fine with Flutter 1.22.4, but fails with later versions.

This [page](https://flutter.dev/docs/release/breaking-changes/network-policy-ios-android) mentions settings (Info plist or manifest) that can be changed to allow insecure connections. However, none of those are needed for our app on iOS/Android versions more recent than those mentioned, and it can still connect over http to a local IP. So I don't think Flutter should be more strict than the platforms.

In our case, we want to connect to an older IoT device that doesn't support HTTPS. Setting `NSAllowsArbitraryLoads` seems too much for that. (We don't want to allow arbitrary connections to the internet.) We can also not whitelist the devices IP because it is dynamically assigned.

Related but closed by the author: https://github.com/flutter/flutter/issues/68550

Minimal example. This assumes a Philips Hue Bridge is available at 192.168.1.21:
```dart
import 'dart:io';

import 'package:flutter/widgets.dart';

void main() async {
  final client = HttpClient();
  
  final request = await client.getUrl(Uri.parse("http://192.168.1.21/api/config"));
  request.close();
  final response = await request.done;
  print(response.statusCode);

  runApp(SizedBox());
}
```

**Expected outcome**: It prints 200. (Or perhaps `Unhandled Exception: SocketException: Connection failed (OS Error: Host is down, errno = 64), address = 192.168.1.217, port = 80` when no device is present with that IP.)

**Actual outcome**: `Unhandled Exception: Bad state: Insecure HTTP is not allowed by platform: http://192.168.1.21/api/config`

Works with:
```
[✓] Flutter (Channel unknown, 1.22.4, on Mac OS X 10.15.7 19H15 darwin-x64, locale en-NL)
    • Flutter version 1.22.4 at ...
    • Framework revision 1aafb3a8b9 (6 weeks ago), 2020-11-13 09:59:28 -0800
    • Engine revision 2c956a31c0
    • Dart version 2.10.4

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.0.rc.1
```
and
```
[✓] Flutter (Channel unknown, 1.22.5, on Mac OS X 10.15.7 19H15 darwin-x64, locale en-NL)
    • Flutter version 1.22.5 at ...
    • Framework revision 7891006299 (12 days ago), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4
```

Using: `IPHONEOS_DEPLOYMENT_TARGET = 9.0;` (the default) or `IPHONEOS_DEPLOYMENT_TARGET = 12.0`.

Fails with:
```
[✓] Flutter (Channel unknown, 1.23.0-18.1.pre, on Mac OS X 10.15.7 19H15 x86_64, locale en-NL)
    • Flutter version 1.23.0-18.1.pre at ...
    • Framework revision 198df796aa (10 weeks ago), 2020-10-15 12:04:33 -0700
    • Engine revision 1d12d82d9c
    • Dart version 2.11.0 (build 2.11.0-213.1.beta)
```
and
```
[✓] Flutter (Channel unknown, 1.24.0-10.2.pre, on Mac OS X 10.15.7 19H15 darwin-x64, locale en-NL)
    • Flutter version 1.24.0-10.2.pre at ...
    • Framework revision 022b333a08 (5 weeks ago), 2020-11-18 11:35:09 -0800
    • Engine revision 07c1eed46b
    • Dart version 2.12.0 (build 2.12.0-29.10.beta)
```
and
```
[✓] Flutter (Channel unknown, 1.25.0-8.1.pre, on Mac OS X 10.15.7 19H15 darwin-x64, locale en-NL)
    • Flutter version 1.25.0-8.1.pre at ...
    • Framework revision 8f89f6505b (6 days ago), 2020-12-15 15:07:52 -0800
    • Engine revision 92ae191c17
    • Dart version 2.12.0 (build 2.12.0-133.2.beta)
```

**Comment by TahaTesser** [comment]
Hi @spkersten 
Can you please provide your `flutter doctor -v` your `flutter run --verbose` and a complete reproducible minimal code sample
Thank you

**Comment by spkersten** [comment]
@TahaTesser I have updated the original comment.

**Comment by spkersten** [comment]
@TahaTesser If I understand the labels, the "found in release: 1.22" isn't needed here, as it works as expected in 1.22.

**Comment by TahaTesser** [comment]
@spkersten 
Thanks! Added by mistake, I was constantly thinking about 1.22 and muscle memory )

**Comment by stuz5000** [comment]
Related https://github.com/flutter/flutter/issues/66275
and closed because 'working as expected'.

Perhaps it's not obvious, but I'll point out that many people build backend in tandem with front-end and it's a phenomenal PITA to make the local server a HTTPs, which involves, among other things: 

 - figuring out how to install a SSL certificate
 - switching the service to allow HTTPs connections
 
 ... all of which may not even be required in a production setting if the service sits behind a firewall that manages HTTPs for you.

Please, can we not require HTTPs for local development.

If there's Android configs that allows workaround for development its not obvious what changes are needed.
(Some iOS config is described here: https://github.com/flutter/flutter/issues/68550, although I'm not clear on whether this bug affect all Dart -- it's certainly badly broken for Android for me)

**Comment by stuz5000** [comment]
The NSAllowsArbitraryLoads is a footgun solution that we be enabled for dev, and left in for production by mistake.

A solution should:

 - allow HTTP for dev
 - require (if desired) HTTPs for production

**Comment by spkersten** [comment]
I don't think the solution should only allow http for dev; it should allow the same as the platforms allow, which is (apparently) to allow http for IP on the local network as well. Otherwise it would still require NSAllowsArbitraryLoads for cases that it isn't needed for. That would be hard to justify (and not desired as it lowers security).

**Comment by stuz5000** [comment]
Sure -- is it possible to enable http access for the local network? 
At least for Android, I was able to access the server on my local machine. After some update, that stopped working.

**Comment by stuz5000** [comment]
What's the source of this problem (is it Android Studio?)
While we're waiting for a fix, is there something that can be revert to allow local development.
Combine this bug with very poor debugging and web, and it's VERY difficult to effectively debug applications in tandem with local development.

**Comment by HarukaMa** [comment]
A temporary solution before this is fixed by flutter (or dart? as it seems the related code is in dart library): Follow [this section](https://flutter.dev/docs/release/breaking-changes/network-policy-ios-android#allowing-cleartext-connection-for-debug-builds) and set the needed options on a debug AndroidManifest.xml or Info.plist. The options will only be used on debug builds so release versions will not be affected by reduced security.

**Comment by Hixie** [comment]
This is probably related to our efforts to follow the mobile platforms' rules on disallowing unencrypted connections, but unfortunately the person who was working on that is no longer around. We might need to just remove this code entirely for now.

cc @xster

**Comment by Hixie** [comment]
For anyone blocked on this in production (as opposed to in development), it would be good if you could describe the setup you are using where this is a problem, in case we can find a quick temporary workaround.

**Comment by xster** [comment]
The original issue causing this was https://github.com/flutter/flutter/issues/54448

**Comment by spkersten** [comment]
@Hixie I believe I've described a setup in the original post. Our use case is that we use older devices in the field that don't support HTTPS from which we still want to get some minimal amount of data. Let me know if you miss something.

If I do the same natively in Swift, for example:
```
        let url = URL(string: "http://192.168.1.21/api/config")!

        let task = URLSession.shared.dataTask(with: url) {(data, _, error) in
            error.map {
                print($0.localizedDescription)
            }
            
            if let data = data {
                print(String(data: data, encoding: .utf8)!)
            }
        }

        task.resume()
```
the request succeeds, while if I try to request `http://www.google.com` (or any other non-HTTPS URL) it fails with (as expected):
```
Task <24EAE187-B531-4977-BA29-6BE51FF48DEC>.<27> finished with error [-1022] Error Domain=NSURLErrorDomain Code=-1022 "The resource could not be loaded because the App Transport Security policy requires the use of a secure connection." UserInfo={NSLocalizedDescription=The resource could not be loaded because the App Transport Security policy requires the use of a secure connection., NSErrorFailingURLStringKey=http://www.google.com/, NSErrorFailingURLKey=http://www.google.com/, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <24EAE187-B531-4977-BA29-6BE51FF48DEC>.<27>"
), _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <24EAE187-B531-4977-BA29-6BE51FF48DEC>.<27>, NSUnderlyingError=0x6000006be400 {Error Domain=kCFErrorDomainCFNetwork Code=-1022 "(null)"}}
```

**Comment by Hixie** [comment]
@spkersten Sorry I meant could you give a little more details? e.g. what language is that "server" side implemented in, what drove you to use local HTTP, that kind of thing.

fyi @lrhn @athomas as I understand it you've been involved in this work, so this may be of interest.

**Comment by spkersten** [comment]
@Hixie The servers are an older version of our "home automation hubs". I wasn't around when they were introduced so I don't know how decisions were made. I'm guessing there might have been memory constraints.

**Comment by Hixie** [comment]
I see, so this is some opaque blob of code that listens for localhost-only connections on the same device. Got it, thanks.

**Comment by spkersten** [comment]
Connections on the same local network.

**Comment by Hixie** [comment]
Oh, I misunderstood. I thought by "local IP" this meant localhost.

Doing unencrypted traffic on the local network seems dubious from a security perspective. What if the local network is a wifi guest network, say? That said, it's not my place to critique your use case.

I would just flip the flag to allow unencrypted connections. I understand that Flutter's policy here is stricter than iOS and Android's, but that doesn't seem unreasonable (e.g. because of the open wifi case). It doesn't make things any less secure to flip the flag, it just means you won't have insecure behaviour flagged. But in your case that seems to be what you need.

**Comment by spkersten** [comment]
Flipping the switch does make things less secure because then any connection the app makes (and it make several to servers on the internet as well) might be (accidentally) insecure. 

Also, I believe that for App Store review, we'd have to give a reason for allowing insecure connections. And we can't because that setting isn't needed, given the default exception for IPs on the local network.

**Comment by Hixie** [comment]
> Flipping the switch does make things less secure because then any connection the app makes (and it make several to servers on the internet as well) might be (accidentally) insecure.

I mean, it's not setting the flag that would make it less secure, it's accidentally making the insecure connection. But sure.

> Also, I believe that for App Store review, we'd have to give a reason for allowing insecure connections. And we can't because that setting isn't needed, given the default exception for IPs on the local network.

This on the other hand is a fair point.

**Comment by Hixie** [comment]
Do we know what iOS uses as its definition of "local network"? Does it look at the network interface's netmask?

**Comment by xster** [comment]
Update: will take a look this week

**Comment by xster** [comment]
Reading through the original discussions at https://github.com/dart-lang/sdk/issues/40548 a bit, I believe unfortunately there's no other solution than to undo the original implementations listed in https://github.com/flutter/flutter/issues/54448#issuecomment-801186901. And unfortunately, I don't see a good way for these reverts to appear before the next stable release. 

I would have liked to suggest using https://developer.apple.com/documentation/bundleresources/information_property_list/nsapptransportsecurity/nsallowslocalnetworking?language=objc except not all info.plist configurations work, and only a hardcoded subset of keys are translated into dart:io https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm#L137. 

As another mental-gymnastic-y way of fitting into NSExceptionDomains which is read, you may perhaps try 

```xml
<dict>
    <key>NSExceptionDomains</key>
    <dict>
        <key>localhost</key>
        <dict>
            <key>NSIncludesSubdomains</key>
            <true/>
            <key>NSExceptionAllowsInsecureHTTPLoads</key>
            <true/>
        </dict>
    </dict>
</dict>
```

but it's a bit of a hail mary. I don't have an easy way to reproduce this locally so I can't test this. Nor do I have an intuition of how Apple review will react to this. 

Barring those, I'm afraid the best we can offer is waiting for the next stable as I revert this mechanism.

**Comment by xster** [comment]
https://dart-review.googlesource.com/c/sdk/+/192953

**Comment by zanderso** [comment]
When the revert lands, it looks like we'll need to re-open https://github.com/flutter/flutter/issues/54448, and ask Dart folks for help over there.

**Comment by xster** [comment]
Note to self to test when both https://dart-review.googlesource.com/c/sdk/+/192953 and https://github.com/flutter/engine/pull/25299 roll

**Comment by chinmaygarde** [comment]
@xster That seems to have been rolled in. Were you able to verify per your last comment?

**Comment by xster** [comment]
I'm still waiting for someone with edit rights to help me merge https://dart-review.googlesource.com/c/sdk/+/192953

**Comment by xster** [comment]
Unfortunately a bit more involved revert since the reverted functions are already used in Google code in various places.

**Comment by mraleph** [comment]
If you are removing public interface members it's a breaking change and might be hard to roll in. (and it needs to follow breaking change process for Dart). I suggest you don't remove public member and instead stub it to return true all the time.

I also would like to understand how did we actually break something. I thought the code which blocked connections was already removed and instead we required manual checking?

**Comment by chinmaygarde** [comment]
The internal patch has been LGTM'ed but not landed. @xster Can you shepherd the patch please?

**Comment by xster** [comment]
Stubbed the dart:io API being reverted. It's still breaking some google tests though at a glance, they don't look related. Debugging.

**Comment by xster** [comment]
There were fields private to dart but public to c++ via vm-entrypoints in https://github.com/flutter/engine/pull/20218/files#diff-26a49cb762edd726727368ff7449a99076c16c642ad99d632466977b8d674bcfR31 that are removed in https://dart-review.googlesource.com/c/sdk/+/192953/8/sdk/lib/io/embedder_config.dart.

**Comment by chinmaygarde** [comment]
@xster You patch has been LGTM'ed and is waiting to land. Are you planning on working on this soon? OTOH, if this issue needs to be reprioritized, please update the priorities. Thansk.

**Comment by xster** [comment]
Now that both https://github.com/dart-lang/sdk/commit/d9683045704cb533cd079f6caf43fdcb3253aebc and https://github.com/flutter/engine/pull/25299 are in the framework, I tested on Android and iOS and local HTTP requests both work. I'm closing this issue. https://github.com/flutter/engine/pull/25623 is an additional nice-to-have that isn't blocking this bug.

**Comment by aam** [comment]
> Reading through the original discussions at [dart-lang/sdk#40548](https://github.com/dart-lang/sdk/issues/40548) a bit, I believe unfortunately there's no other solution than to undo the original implementations listed in [#54448 (comment)](https://github.com/flutter/flutter/issues/54448#issuecomment-801186901). And unfortunately, I don't see a good way for these reverts to appear before the next stable release.

Did we consider use of zone `dart.library.io.allow_http` for the client code that wants to access http (https://github.com/dart-lang/sdk/blob/master/tests/standalone/io/http_ban_http_embedder_test.dart#L27):

```
    runZoned(() => httpClient.getUrl(httpUri),
        zoneValues: {#dart.library.io.allow_http: true});
```

**Comment by aam** [comment]
Ah, never mind. I see that https://dart-review.googlesource.com/c/sdk/+/154180 that was actually reverted removed use of zone variable `dart.library.io.allow_http` leaving only dart environment variables `dart.library.io.may_insecurely_connect_to_all_domains` and `dart.library.io.domain_network_policies` in control, which don't seem to be writable by flutter client application.

**Comment by aam** [comment]
Actually, reverted change https://dart-review.googlesource.com/c/sdk/+/154180 that introduced  domain_network_policies also provided for user to provide application-level overrides via custom section in AndroidManifest.xml/Info.plist. Has this been tried?
From very first comment https://github.com/flutter/flutter/issues/72723#issue-772290171 I believe https://flutter.dev/docs/release/breaking-changes/network-policy-ios-android#migration-guide was confused with next section that talks about `Allowing cleartext connection for debug builds`, which seems to be a different solution to somewhat similar problem.
